suplemon/prompt.py

```diff
@@ -0,0 +1,251 @@
+# -*- encoding: utf-8
+"""
+Promts based on the Editor class for querying user input.
+"""
+
+import os
+
+from .editor import Editor
+
+
+class Prompt(Editor):
+    """An input prompt based on the Editor."""
+    def __init__(self, app, window):
+        Editor.__init__(self, app, window)
+        self.ready = 0
+        self.canceled = 0
+        self.input_func = lambda: False
+        self.caption = ""
+
+    def init(self):
+        Editor.init(self)
+        # Remove the find feature, otherwise it can be invoked recursively
+        del self.operations["find"]
+
+    def set_config(self, config):
+        """Set the configuration for the editor."""
+        # Override showing line numbers
+        config["show_line_nums"] = False
+        Editor.set_config(self, config)
+
+    def set_input_source(self, input_func):
+        # Set the input function to use while looping for input
+        self.input_func = input_func
+
+    def on_ready(self):
+        """Accepts the current input."""
+        self.ready = 1
+        return
+
+    def on_cancel(self):
+        """Cancels the input prompt."""
+        self.set_data("")
+        self.ready = 1
+        self.canceled = 1
+        return
+
+    def line_offset(self):
+        """Get the x coordinate of beginning of line."""
+        return len(self.caption)+1
+
+    def render_line_contents(self, line, pos, x_offset, max_len):
+        """Render the prompt line."""
+        x_offset = self.line_offset()
+        # Render the caption
+        self.window.addstr(pos[1], 0, self.caption)
+        # Render input
+        self.render_line_normal(line, pos, x_offset, max_len)
+
+    def handle_input(self, event):
+        """Handle special bindings for the prompt."""
+        name = event.key_name
+        if name in ["ctrl+c", "escape"]:
```

Comment:
Well, the idea for `key_mappings.py` was to be able to use the same format for key config as Sublime Text does. Even though `curses` has lots of definitions they're a bit cryptic and they vary slightly between Python 2 and 3. Somehow it just feels natural to use the readable versions in the code too.

I think I'll go with your second solution regarding the modifier ordering. Seems the most straight forward way to do it.